File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] before = new int[nums.length];
        int[] result = new int[nums.length];
        before[0] = 1;
        for(int i = 1; i < nums.length; i++) {
            before[i] = nums[i - 1] * before[i - 1];
        }

        result[nums.length - 1] = before[nums.length - 1];
        for(int j = nums.length - 2; j >= 0; j--) {
            result[j] = before[j] * nums[j + 1];
            nums[j] *= nums[ j + 1];
        }

        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> before(nums.size(), 1);
        vector<int> after(nums.size(), 1);
        vector<int> res(nums.size());

        for (auto i = 1uz; i < nums.size(); ++i) {
            before[i] = nums[i - 1] * before[i - 1];
        }

        for (auto i = nums.size() - 1; i-- > 0;) {
            after[i] = nums[i + 1] * after[i + 1];
        }

        for (auto i = 0uz; i < res.size(); ++i) {
            res[i] = before[i] * after[i];
        }

        return res;
    }
};